fix: sampler docs (#125)
scverse/annbatch: docs/custom-sampler.md

# Advanced: Implementing a Custom Sampler

## Overview

To implement a custom sampler, you need to understand two key components:

### 1. `annbatch.abc.Sampler`

> This is the abstract base class that all samplers must inherit from. You need to implement:

Author: I am not aware of a way to embed docstrings in other pages, but it might be possible.  I think it would look strange though.

import numpy as np

class RandomSampler(Sampler):

    def __init__(self, n_obs: int, batch_size: int):
        self.n_obs = n_obs
        self.batch_size = batch_size

    def _sample(self, n_obs: int) -> Iterator[LoadRequest]:
        for i in np.array_split(np.random.default_rng().permutation(self.n_obs), self.n_obs // self.batch_size):
            yield {"splits": [np.arange(self.batch_size)], "chunks": [slice(idx, idx + 1) for idx in i]}

```

## Performance Considerations

When implementing a custom sampler, it's crucial to consider **disk access patterns** to ensure efficient data loading. The performance of your sampler heavily depends on how it accesses data from disk-backed arrays.

### Efficient vs. Inefficient Access Patterns

**✅ Example 1 (ChunkedSampler) - Efficient:**
- Loads **contiguous chunks** of data: `[slice(0,100), slice(100,200), slice(200,300)]`
- Minimizes disk seeks by reading sequential blocks
- Takes advantage of chunk-based storage formats
- Optimal for Zarr arrays where data is stored in chunks

```
Disk access pattern (sequential):
Read chunk 0 → Read chunk 1 → Read chunk 2 → ...
└─────────┴─────────┴─────────┘
     Fast sequential reads
```

> **❌ Example 2 (RandomSampler) - Inefficient:**

Reviewer: I've rewritten this section, I think the key takeaways for the user should be:
* Chunked reads over random reads
* Chunked reads come with a trade-off + how to mitigate this trade-off

Let me know what you think

# Advanced: Implementing a Custom Sampler

## Overview

To implement a custom sampler, you need to understand two key components:

> ### 1. {class}`annbatch.abc.Sampler`

Author: ```suggestion
### {class}`annbatch.abc.Sampler`
```

This is the abstract base class that all samplers must inherit from. You need to implement:

> - **{meth}`annbatch.abc.Sampler._sample`**: The core sampling logic that generates load requests. This method receives the total number of observations and yields {class}`annbatch.types.LoadRequest` that specify how data should be loaded and batched. See below for more information. The outer {meth}`annbatch.abc.Sampler.sample` that wraps this implemented method calls {meth}`annbatch.abc.Sampler.validate` at runtime.

Author: ```suggestion
- **{meth}`~annbatch.abc.Sampler._sample`**: The core sampling logic that generates load requests. This method receives the total number of observations and yields {class}`annbatch.types.LoadRequest` that specify how data should be loaded and batched. See below for more information. The outer {meth}`annbatch.abc.Sampler.sample` that wraps this implemented method calls {meth}`annbatch.abc.Sampler.validate` at runtime.
```

# Advanced: Implementing a Custom Sampler

## Overview

To implement a custom sampler, you need to understand two key components:

### 1. `annbatch.abc.Sampler`

This is the abstract base class that all samplers must inherit from. You need to implement:

- **{meth}`annbatch.abc.Sampler._sample`**: The core sampling logic that generates load requests. This method receives the total number of observations and yields {class}`annbatch.types.LoadRequest that specify how data should be loaded and batched.  See below for more information.  The outer {meth}`annbatch.abc.Sampler.sample` that wraps this implemented method calls {meth}`annbatch.abc.Sampler.validate` at runtime.

- **`validate(n_obs: int) -> None`** (optional): Validates the sampler configuration against the given number of observations. Override this method to add custom validation for your sampler parameters. It should raise a `ValueError` if the configuration is invalid.

### 2. `annbatch.types.LoadRequest`

A `TypedDict` that specifies how data should be loaded. Each `LoadRequest` contains:

> - **`chunks`**: A list of slices that define which chunks to load from disk. Each slice should have a range up to the chunk_size (except the last one, which may be smaller but not empty). These slices determine which portions of the dataset are read into memory.

Author: ```suggestion
- **`chunks`**: A list of slices that define which contiguous chunks of memory to load from disk. Each slice should have a range up to the `chunk_size` (except the last one, which may be smaller but not empty). These slices determine which portions of the dataset are read into memory.
```

  ```
  Example: Loading random chunks from a large array

  Full collection (virtual conncatentation of all on disk files) (e.g., 1000 observations, chunk_size=100):
  ┌─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
  │ Chunk 0 │ Chunk 1 │ Chunk 2 │ Chunk 3 │ Chunk 4 │ Chunk 5 │ Chunk 6 │ Chunk 7 │ Chunk 8 │ Chunk 9 │
  │  0-99   │ 100-199 │ 200-299 │ 300-399 │ 400-499 │ 500-599 │ 600-699 │ 700-799 │ 800-899 │ 900-999 │
  └─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

>   LoadRequest with chunks = [slice(200,300), slice(700,800), slice(0,100), slice(500,600)]:

Author: I think these should be unaligned still, which I am now more convinced of because it's a virtual concatenation of datasets, not zarr chunks (see below)

  │  0-99   │         │ 200-299 │         │         │ 500-599 │         │ 700-799 │         │         │
  └─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘
      ↓                   ↓                             ↓                     ↓
      └───────────────────┴─────────────────────────────┴─────────────────────┘
                                          ↓
  Loaded into memory and concatenated (400 observations):
  ┌─────────┬─────────┬─────────┬─────────┐
  │ 200-299 │ 700-799 │  0-99   │ 500-599 │
  └─────────┴─────────┴─────────┴─────────┘
  ```
  **Important:** The number of samples that get loaded into memory at once, should be devisible by the batch size. Otherwise, the reminder will get dropped.

- **`splits`** (optional): A list of numpy arrays that define how the loaded data should be split into batches after being read from disk and concatenated in memory.
  - If not supplied: batches are randomly created based on the loaded chunks.
  - If supplied: you can control how batches are created from the in-memory chunks. Each array contains indices that map into the concatenated in-memory data.

>   - The `splits` parameter gives you fine-grained control over how individual batches are created based on the loaded chunks. This is particularly useful when you want to organize batches based on semantic labels, categories, or other metadata.

Author: ```suggestion
- **`splits`** (optional): A list of integer {class}`numpy.ndarray`s that define how the loaded data should be split into batches after being read from disk and concatenated in memory. Each array contains indices that map into the concatenated in-memory data. This control is particularly useful when you want to organize batches based on semantic labels, categories, or other metadata.
  - *If not supplied*: batches are randomly created based on the loaded chunks.
```

            batch_indices = [
                np.arange(i, min(i + self.batch_size, chunk_size_actual))
                for i in range(0, chunk_size_actual, self.batch_size)
            ]
            yield {"chunks": [chunk], "splits": batch_indices}
```

## Example 2: Implementing a `RandomSampler` class
```python
from annbatch.abc import Sampler
from collections.abc import Iterator
from annbatch.types import LoadRequest
import numpy as np

class RandomSampler(Sampler):

    def __init__(self, n_obs: int, batch_size: int):
        self.n_obs = n_obs
        self.batch_size = batch_size

    def _sample(self, n_obs: int) -> Iterator[LoadRequest]:
        for i in np.array_split(np.random.default_rng().permutation(self.n_obs), self.n_obs // self.batch_size):
            yield {"splits": [np.arange(self.batch_size)], "chunks": [slice(idx, idx + 1) for idx in i]}

```

## Performance Considerations

When implementing a custom sampler, it's crucial to consider **disk access patterns** to ensure efficient data loading.
The performance of your sampler heavily depends on how it accesses data from disk-backed arrays.

### The Core Strategy: Blocked Reads + In-Memory Shuffling

> The key to efficient sampling from disk-backed arrays is to **read data in contiguous blocks** and then **shuffle in memory**.

Author: ```suggestion
The key to efficient sampling from disk-backed arrays is to **read data from many contiguous blocks** (i.e., more than your batch size generally) and then **shuffle in memory**.
```
